firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,499 @@
+        """Get the customized ABI for the update service.
+
+        :returns: ABI version
+        """
+        abi = self.app_info.XPCOMABI
+        if mozinfo.isMac:
+            abi += self.marionette.execute_script("""
+              let macutils = Cc['@mozilla.org/xpcom/mac-utils;1']
+                             .getService(Components.interfaces.nsIMacUtils);
+              if (macutils.isUniversalBinary) {
+                return '-u-' + macutils.architecturesInBinary;
+              }
+              return '';
+            """)
+
+        return abi
+
+    @property
+    def active_update(self):
+        """ Holds a reference to an ActiveUpdate object."""
+        return self._active_update
+
+    @property
+    def allowed(self):
+        """Check if the user has permissions to run the software update
+
+        :returns: Status if the user has the permissions
+        """
+        allowed = self.marionette.execute_script("""
+          let aus = Cc['@mozilla.org/updates/update-service;1']
+                    .getService(Components.interfaces.nsIApplicationUpdateService);
+          return aus.canCheckForUpdates && aus.canApplyUpdates;
+        """)
+
+        return bool(allowed)
+
+    @property
+    def build_info(self):
+        """Return information of the current build version
+
+        :returns: A dictionary of build information
+        """
+        return {
+            'buildid': self.app_info.appBuildID,
+            'channel': self.update_channel.channel,
+            'disabled_addons': self.prefs.get_pref(self.PREF_DISABLED_ADDONS),
+            'locale': self.app_info.locale,
+            'mar_channels': self.mar_channels.channels,
+            'url_aus': self.get_update_url(True),
+            'user_agent': self.app_info.user_agent,
+            'version': self.app_info.version
+        }
+
+    @property
+    def is_complete_update(self):
+        """Return true if the offered update is a complete update
+
+        :returns: True if the offered update is a complete update
+        """
+        # Throw when isCompleteUpdate is called without an update. This should
+        # never happen except if the test is incorrectly written.
+        assert self.active_update.exists, 'An active update has been found'
+
+        patch_count = self.active_update.patch_count
+        assert patch_count == 1 or patch_count == 2,\
+            'An update must have one or two patches included'
+
+        # Ensure Partial and Complete patches produced have unique urls
+        if patch_count == 2:
+            patch0_URL = self.active_update.get_patch_at(0)['URL']
+            patch1_URL = self.active_update.get_patch_at(1)['URL']
+            assert patch0_URL != patch1_URL,\
+                'Partial and Complete download URLs are different'
+
+        return self.active_update.selected_patch['type'] == 'complete'
+
+    @property
+    def mar_channels(self):
+        """ Holds a reference to a MARChannels object."""
```

Comment:
nit: you may want to linkify the class name like ":class:MARChannels". This might also apply to some other methods.